rust/src/events/internal_metadata.rs

```diff
@@ -0,0 +1,416 @@
+ *
+ */
+
+//! Implements the internal metadata class attached to events.
+//!
+//! The internal metadata is a bit like a `TypedDict`, in that it is stored as a
+//! JSON dict in the DB. Most events have zero, or only a few, of these keys
+//! set. Therefore, since we care more about memory size than performance here,
+//! we store these fields in a mapping.
+//!
+//! We want to store (most) of the fields as Rust objects, so we implement the
+//! mapping by using a vec of enums. This is less efficient than using
+//! attributes, but for small number of keys is actually faster than using a
+//! hash or btree map.
+
+use std::{num::NonZeroI64, ops::Deref};
+
+use anyhow::Context;
+use log::warn;
+use pyo3::{
+    exceptions::PyAttributeError, pyclass, pymethods, types::PyDict, IntoPy, PyAny, PyObject,
+    PyResult, Python,
+};
+
+/// Definitions of the various fields of the internal metadata.
+#[derive(Clone)]
+enum EventInternalMetadataData {
+    OutOfBandMembership(bool),
+    SendOnBehalfOf(Box<str>),
+    RecheckRedaction(bool),
+    SoftFailed(bool),
+    ProactivelySend(bool),
+    Redacted(bool),
+    TxnId(Box<str>),
+    TokenId(i64),
+    DeviceId(Box<str>),
+}
+
+impl EventInternalMetadataData {
+    /// Convert the field to its name and python object.
+    fn to_python_pair(&self, py: Python<'_>) -> (&'static str, PyObject) {
+        match self {
+            EventInternalMetadataData::OutOfBandMembership(o) => {
+                ("out_of_band_membership", o.into_py(py))
+            }
+            EventInternalMetadataData::SendOnBehalfOf(o) => ("send_on_behalf_of", o.into_py(py)),
+            EventInternalMetadataData::RecheckRedaction(o) => ("recheck_redaction", o.into_py(py)),
+            EventInternalMetadataData::SoftFailed(o) => ("soft_failed", o.into_py(py)),
+            EventInternalMetadataData::ProactivelySend(o) => ("proactively_send", o.into_py(py)),
+            EventInternalMetadataData::Redacted(o) => ("redacted", o.into_py(py)),
+            EventInternalMetadataData::TxnId(o) => ("txn_id", o.into_py(py)),
+            EventInternalMetadataData::TokenId(o) => ("token_id", o.into_py(py)),
+            EventInternalMetadataData::DeviceId(o) => ("device_id", o.into_py(py)),
+        }
+    }
+
+    /// Converts from python key/values to the field.
+    ///
+    /// Returns `None` if the key is a valid but unrecognized string.
+    fn from_python_pair(key: &PyAny, value: &PyAny) -> PyResult<Option<Self>> {
+        let key_str: &str = key.extract()?;
+
+        let e = match key_str {
+            "out_of_band_membership" => EventInternalMetadataData::OutOfBandMembership(
+                value
+                    .extract()
+                    .with_context(|| format!("'{key_str}' has invalid type"))?,
+            ),
+
+            "send_on_behalf_of" => EventInternalMetadataData::SendOnBehalfOf(
+                value
+                    .extract::<String>()
+                    .map(String::into_boxed_str)
```

Comment:
Because you're mapping with the explicit type after, I think the generic parameter can be inferred

```suggestion
                    .extract()
                    .map(String::into_boxed_str)
```